### PR TITLE
STORM-2720 : Add TIMESTAMP option for FirstPollOffset for Kafka Trident spout

### PR DESCRIPTION
**Pull request by janithkv**

### CODE REVIEW

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java [hunk 1]
```diff
@@ -137,7 +142,11 @@ public KafkaSpoutConfig(Builder<K, V> builder) {
         /**
          * The kafka spout polls records from the last committed offset, if any. If no offset has been committed it behaves as LATEST
          */
-        UNCOMMITTED_LATEST;
+        UNCOMMITTED_LATEST,
+        /**
+         * Start at the earliest offset whose timestamp is greater than or equal to the given startTimestamp
+         */
+        TIMESTAMP;
```
[the remark this answers is not in the document]
Comment (janithkv) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java]: Hmm not sure. I think practically it might make sense to go back a few minutes rather than all the way upto EARLIEST. But that might be a bit out of scope.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java [hunk 2]
```diff
@@ -225,7 +229,23 @@ private void emitTuple(TridentCollector collector, ConsumerRecord<K, V> record)
      */
     private long seek(TopicPartition tp, KafkaTridentSpoutBatchMetadata lastBatchMeta) {
         if (isFirstPoll(tp)) {
-            if (firstPollOffsetStrategy == EARLIEST) {
+            if(firstPollOffsetStrategy.equals(TIMESTAMP)) {
+                Long startTimeStampOffset = null;
+                try {
+                    startTimeStampOffset =
+                            consumer.offsetsForTimes(Collections.singletonMap(tp, startTimeStamp)).get(tp).offset();
+                } catch (IllegalArgumentException e) {
```
[the remark this answers is not in the document]
Comment (janithkv) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java]: ok will change

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java [hunk 3]
```diff
@@ -225,7 +229,23 @@ private void emitTuple(TridentCollector collector, ConsumerRecord<K, V> record)
...
+                    LOG.error("Illegal timestamp {} provided for tp {} ",startTimeStamp,tp.toString());
+                } catch (UnsupportedVersionException e) {
+                    LOG.error("Kafka Server do not support offsetsForTimes(), probably < 0.10.1",e);
+                }
+                if(startTimeStampOffset != null) {
+                    LOG.debug("First poll for topic partition [{}], seeking to partition from startTimeStamp [{}]", tp , startTimeStamp);
+                    consumer.seek(tp, startTimeStampOffset);
+                } else {
+                    LOG.info("Kafka consumer offset reset by timestamp failed for TopicPartition {}, TimeStamp {}, Offset {}. Restart with a different Strategy ", tp, startTimeStamp, startTimeStampOffset);
```
[the remark this answers is not in the document]
Comment (janithkv) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java]: ok will change

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (srdo) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java]: Not sure I follow. I mean that I think we should have both TIMESTAMP, which unconditionally seeks to a timestamp, and UNCOMMITTED_TIMESTAMP which only seeks to a timestamp if no offset has been committed yet. That way you can use UNCOMMITTED_TIMESTAMP to start at a timestamp the first time you deploy the topology, but the topology doesn't reset when you redeploy.

**Comment by HeartSaVioR** [comment]
This is definitely a good feature to add. @janithkv Do you have any update? I guess you didn't actually address @srdo comments. Please let me know if you couldn't continue working on this - I'll see if I can continuous working on top of this patch. Thanks in advance!

**Comment by janithkv** [comment]
@HeartSaVioR @srdo , This totally went out of my radar. Fixing it right now

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (janithkv) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java]: Agreed. Adding that.

**Comment by janithkv** [comment]
@srdo @HeartSaVioR , Could you guys take a look.
1. Not sure how to make the PR build
2. Any suggestions on how to add a test for TIMESTAMP case ? MockConsumer does not seem to have offsetsForTimes implemented.

**Comment by srdo** [comment]
1. The build is broken due to https://github.com/apache/storm/pull/3007. Don't worry about it, once that one gets merged, you can just rebase this and it should work.

2. Yes, as you note MockConsumer doesn't have this support implemented. You could use a Mockito spy instead https://static.javadoc.io/org.mockito/mockito-core/2.23.0/org/mockito/Mockito.html#16.

**Comment by janithkv** [comment]
@srdo , will try to use Mockito, Also could you check if I addressed everything you wanted ?

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java [hunk 4]
```diff
@@ -279,6 +293,25 @@ private long seek(TopicPartition tp, KafkaTridentSpoutBatchMetadata lastBatchMet
...
     }
 
+    /**
+     * Seek the consumer to offset corresponding to startTimeStamp.
+     */
+    private void seekOffsetByStartTimeStamp(TopicPartition tp) {
+        long startTimeStampOffset;
+        try {
+            startTimeStampOffset =
+                    consumer.offsetsForTimes(Collections.singletonMap(tp, startTimeStamp)).get(tp).offset();
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java]: Nit: I would like this broken up into multiple statements so we can tell where it breaks if we get an NPE.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java [hunk 5]
```diff
@@ -279,6 +293,25 @@ private long seek(TopicPartition tp, KafkaTridentSpoutBatchMetadata lastBatchMet
...
+        } catch (IllegalArgumentException e) {
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java]: Why do we need to catch and handle these exceptions? Can't we just not catch them and it'll work?

**Comment by srdo** [comment]
Yes, I think you did. Left a few new comments.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java (hunk 5 above)
Comment (janithkv) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java]: I was thinking , it would come up better in the UI, with ReportedFailedExceptions. But I can take it off, if it does not add value.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java (hunk 4 above)
Comment (janithkv) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java]: ok

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java (hunk 5 above)
Comment (janithkv) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java]: Actually at this point, I guess, just letting the exceptions go, without catching should be fine. Removing the catch statements

**Comment by janithkv** [comment]
@srdo , comments addressed

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java [hunk 6]
```diff
@@ -279,6 +293,18 @@ private long seek(TopicPartition tp, KafkaTridentSpoutBatchMetadata lastBatchMet
         return fetchOffset;
     }
 
+    /**
+     * Seek the consumer to offset corresponding to startTimeStamp.
+     */
+    private void seekOffsetByStartTimeStamp(TopicPartition tp) {
+        long startTimeStampOffset;
+        Map<TopicPartition, OffsetAndTimestamp> map = consumer.offsetsForTimes(Collections.singletonMap(tp, startTimeStamp));
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java]: Nit: Not a very useful variable name. How about `offsetsForTimes`?

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java [hunk 7]
```diff
@@ -303,4 +315,38 @@ public void testLatestStrategyWhenTopologyIsRedeployed() {
...
+        long preRestartEmittedOffset = 20;
+        int preRestartEmittedRecords = 10;
+        long timeStampStartOffset = 2L;
+        long pollTimeout = 1L;
+        KafkaTridentSpoutBatchMetadata preExecutorRestartLastMeta = new KafkaTridentSpoutBatchMetadata(preRestartEmittedOffset, preRestartEmittedOffset + preRestartEmittedRecords - 1, "Some older topology");
+
+        Consumer kafkaConsumer = Mockito.mock(Consumer.class);
+        when(kafkaConsumer.assignment()).thenReturn(Collections.singleton(partition));
+        OffsetAndTimestamp offsetAndTimestamp = new OffsetAndTimestamp(timeStampStartOffset, startTimeStamp);
+        HashedMap map = new HashedMap();
```
Comment (srdo) [inline external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java]: Why `HashedMap` and not a regular `HashMap`?

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java [hunk 8]
```diff
@@ -81,15 +88,20 @@ public void setUp() {
         records.forEach(record -> consumer.addRecord(record));
     }
 
+    private KafkaTridentSpoutEmitter<String, String> createEmitter(Consumer kafkaConsumer, FirstPollOffsetStrategy firstPollOffsetStrategy) {
```
Comment (srdo) [inline external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java]: Raw type. Please add generics to the Consumer.

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java [hunk 9]
```diff
@@ -303,4 +315,38 @@ public void testLatestStrategyWhenTopologyIsRedeployed() {
...
+         * TIMESTAMP strategy should be applied if the emitter is new and the topology has been redeployed (storm id has changed)
+         * Offset should be reset according to the offset corresponding to startTimeStamp
+         */
+        long preRestartEmittedOffset = 20;
+        int preRestartEmittedRecords = 10;
+        long timeStampStartOffset = 2L;
+        long pollTimeout = 1L;
+        KafkaTridentSpoutBatchMetadata preExecutorRestartLastMeta = new KafkaTridentSpoutBatchMetadata(preRestartEmittedOffset, preRestartEmittedOffset + preRestartEmittedRecords - 1, "Some older topology");
+
+        Consumer kafkaConsumer = Mockito.mock(Consumer.class);
```
Comment (srdo) [inline external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java]: This is a nitpick, but consider using partial mocking for this instead (https://static.javadoc.io/org.mockito/mockito-core/2.27.0/org/mockito/Mockito.html#partial_mocks), and wrap a real KafkaConsumer. That way you won't have to specify as much behavior for the consumer, and the test will be easier to maintain.

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java [hunk 10]
```diff
@@ -303,4 +315,38 @@ public void testLatestStrategyWhenTopologyIsRedeployed() {
...
+        HashedMap topicPartitionMap = new HashedMap();
+        List<ConsumerRecord<String, String>> newRecords = SpoutWithMockedConsumerSetupHelper.createRecords(partition, timeStampStartOffset, recordsInKafka);
+        topicPartitionMap.put(partition,newRecords);
+        when(kafkaConsumer.poll(pollTimeout)).thenReturn(new ConsumerRecords<String,String>(topicPartitionMap));
+        KafkaTridentSpoutEmitter<String, String> emitter = createEmitter(kafkaConsumer, FirstPollOffsetStrategy.TIMESTAMP);
+        TransactionAttempt txid = new TransactionAttempt(0L, 0);
+        KafkaTridentSpoutTopicPartition kttp = new KafkaTridentSpoutTopicPartition(partition);
+        Map<String, Object> meta = emitter.emitPartitionBatchNew(txid, collectorMock, kttp, preExecutorRestartLastMeta.toMap());
+
+        verify(collectorMock, times(recordsInKafka)).emit(emitCaptor.capture());
```
Comment (srdo) [inline external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java]: The test doesn't actually show that TIMESTAMP works currently. You've made the consumer return some messages on poll, and now you're validating that those messages are returned when the emitter polls. I think the test would pass even if TIMESTAMP wasn't implemented properly.

I think a better test would be one of the following:
* Modify the current test to verify that consumer.seek was called with the expected offset before poll was called
Or
* Use a partial mock for the consumer, put some real records in Kafka, wait a few milliseconds, take a timestamp, put a few more records in Kafka. Start the emitter at the stored timestamp, and verify that only the second set of messages are received.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java (hunk 6 above)
Comment (janithkv) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java]: Np

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java (hunk 7 above)
Comment (janithkv) [reply external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java]: Fixing

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java (hunk 8 above)
Comment (janithkv) [reply external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java]: ok

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java (hunk 9 above)
Comment (janithkv) [reply external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java]: I would have to replicate the exact same amount with the partial mock right 
1. add assignment or mock assignment
2. offsetsForTimes would have to be mocked anyway
3. add records or mock poll

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java (hunk 10 above)
Comment (janithkv) [reply external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java]: Adding a verify on seek

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java (hunk 9 above)
Comment (srdo) [reply external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java]: Yes, but it would be more robust to change. If there's an unexpected call to the stub (e.g. let's say we start calling `consumer.beginningOffsets` for some reason), the test will start failing. With a partial mock, we don't have to update the stub in that type of case.

The test is fine the way it is, we can replace it later if it becomes an issue.

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java (hunk 10 above)
Comment (srdo) [reply external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java]: Thanks

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/FirstPollOffsetStrategy.java [hunk 11]
```diff
@@ -31,12 +31,21 @@
      * topology deployment
      */
     LATEST,
+    /**
+     * The kafka spout polls records starting at the earliest offset whose timestamp is greater than or equal to the given startTimestamp.
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/FirstPollOffsetStrategy.java]: We should mention that these options only work for the Trident spout. I'd like to see this implemented for the regular spout as well. Are you up for adding support here, or would you rather leave it for someone else to handle later?

If we don't add support here, we need to mention in this javadoc that it doesn't work for the regular spout.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java [hunk 12]
```diff
@@ -369,6 +369,7 @@ public int getMetricsTimeBucketSizeInSecs() {
 
     @Override
     public String toString() {
+
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java]: Nit: Unnecessary whitespace

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java [hunk 13]
```diff
@@ -240,6 +246,7 @@ private void emitTuple(TridentCollector collector, ConsumerRecord<K, V> record)
      * @return the offset of the next fetch
      */
     private long seek(TopicPartition tp, KafkaTridentSpoutBatchMetadata lastBatchMeta) {
+
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java]: Unnecessary whitespace

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java [hunk 14]
```diff
@@ -303,4 +311,41 @@ public void testLatestStrategyWhenTopologyIsRedeployed() {
...
+         * TIMESTAMP strategy should be applied if the emitter is new and the topology has been redeployed (storm id has changed)
+         * Offset should be reset according to the offset corresponding to startTimeStamp
+         */
+        long preRestartEmittedOffset = 20;
+        int preRestartEmittedRecords = 10;
+        long timeStampStartOffset = 2L;
+        long pollTimeout = 1L;
+        KafkaTridentSpoutBatchMetadata preExecutorRestartLastMeta = new KafkaTridentSpoutBatchMetadata(preRestartEmittedOffset, preRestartEmittedOffset + preRestartEmittedRecords - 1, "Some older topology");
+
+
```
Comment (srdo) [inline external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java]: Unnecessary whitespace

**Comment by srdo** [comment]
Thanks @janithkv, the code looks good. I've left a final few comments, but I think this is nearly there. Thank you for your patience.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/FirstPollOffsetStrategy.java (hunk 11 above)
Comment (janithkv) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/FirstPollOffsetStrategy.java]: I was hoping to get this in for the trident spout first. I had delayed it long enough :(. Adding the comment.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java (hunk 12 above)
Comment (janithkv) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java]: removed

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java (hunk 13 above)
Comment (janithkv) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java]: removed

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java (hunk 14 above)
Comment (janithkv) [reply external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java]: Removed the extra.

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java (hunk 9 above)
Comment (janithkv) [reply external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java]: Aha. thanks for the explanation. That does make sense

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java [hunk 15]
```diff
@@ -279,6 +292,18 @@ private long seek(TopicPartition tp, KafkaTridentSpoutBatchMetadata lastBatchMet
         return fetchOffset;
     }
 
+    /**
+     * Seek the consumer to offset corresponding to startTimeStamp.
+     */
+    private void seekOffsetByStartTimeStamp(TopicPartition tp) {
+        long startTimeStampOffset;
```
Comment (srdo) [inline external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java]: Nit: No reason to not declare and assign on the same line.

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java [hunk 16]
```diff
@@ -26,11 +26,12 @@
...
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
-import org.apache.kafka.clients.consumer.ConsumerRecord;
-import org.apache.kafka.clients.consumer.MockConsumer;
-import org.apache.kafka.clients.consumer.OffsetResetStrategy;
+
+import org.apache.commons.collections.map.HashedMap;
+import org.apache.kafka.clients.consumer.*;
```
Comment (srdo) [inline external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java]: Nit: Please avoid wildcard imports, they make the code harder to read

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/FirstPollOffsetStrategy.java (hunk 11 above)
Comment (srdo) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/FirstPollOffsetStrategy.java]: That's fair, just wanted to be sure we noted somewhere that it doesn't work for the regular spout.

File: external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java (hunk 15 above)
Comment (janithkv) [reply external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java]: yes. moved

File: external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java (hunk 16 above)
Comment (janithkv) [reply external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitterEmitTest.java]: 🤦‍♂

**Comment by janithkv** [comment]
@srdo , Do you have a good ref, for squashing to a single commit. I think I messed up my local branch.

**Comment by srdo** [comment]
There's a decent guide here https://stackabuse.com/git-squash-multiple-commits-in-to-one-commit/

Basically you want to do ``git rebase -i `git merge-base HEAD origin/master` `` (where origin/master is Storm's master branch). This will open an editor (usually vim, if you'd rather use nano or something else, do `git config core.editor nano` before doing the rebase command).

In the editor, you'll see a list of your commits, with the "pick" command in front of them. If you replace "pick" with "fixup", you'll merge the line's commit into the previous commit.

What you want is for the editor to show something like
```
pick commit1
fixup commit2
fixup commit3
```

Once you've edited the file, save and quit the editor. Git will execute the rebase and squash your commits down to one.

If you think you've messed up, you can always do `git reflog` to see the recent commits you've been on. If you need to reset back to an old commit, you can do so with `git reset --hard <commit>`. This way, you can always get back to a good commit, even if you've messed something up.

**Comment by HeartSaVioR** [comment]
I'll just handle squashing while merging, as this PR waited more than 6 months.

**Comment by HeartSaVioR** [comment]
I just realized this PR has some conflicts, it doesn't look like easy to fix. Even applying diff via patch file doesn't work. Is this patch be compatible with current master branch? If not, maybe crafting revised patch manually against master branch might be faster.

**Comment by srdo** [comment]
Github doesn't show any conflicts?

**Comment by HeartSaVioR** [comment]
So final diff doesn't have conflict with current master, but even first commit has conflict. I feel it's just easier to play with diff file and provide authorship on new commit. Otherwise @janithkv could craft a new PR against master.

**Comment by srdo** [comment]
I think we can just use the squash-and-merge button in Github? There's an option to squash on merge on the "Merge pull request" button if you click the little arrow.

If not, it is easy to squash without rebase when merging via `git merge --squash`. Here's a branch with this PR squashed https://github.com/srdo/storm/tree/STORM-2720-squash2

**Comment by HeartSaVioR** [comment]
Yeah we could do it and that has been my favorite option of merging for other projects. I just have been thinking some of us have been debated against auto-merge script (1. some of us preferred merge commit, 2. some of us concerned about squashing commits, though the concern was mostly about multi authors).

**Comment by HeartSaVioR** [comment]
Just did it. I think it's time to discuss again for "how to merge" since I think we should have simpler approach to reduce maintain cost.

**Comment by srdo** [comment]
Thanks. I think last time we discussed this we agreed that squashing was fine, unless there were multiple authors, in which case it should be squashed to one commit per author if possible.

So for PRs like this, we can probably just go ahead and use squash and merge in Github IMO. It's easier for contributors as well if we don't ask them to do the squash.

**Comment by HeartSaVioR** [comment]
> It's easier for contributors as well if we don't ask them to do the squash.
That was my major concern while debating - I'm happy to see someone in the same page regarding this. :)

**Comment by Ethanlm** [comment]
I do prefer to use squash-and-merge button. It makes the git history very clean.
